### PR TITLE
Added a single color <paper-spinner-lite>

### PR DESCRIPTION
**Pull request by keanulee**
This branch adds a single color `<paper-spinner-lite>` element that uses less DOM elements. This is much cheaper than styling all 4 layers with the same color, and this should improve performance.

### CODE REVIEW

File: demo/index.html [hunk 1]
```diff
@@ -16,45 +16,37 @@
...
     <link rel="import" href="../paper-spinner.html">
+    <link rel="import" href="../paper-spinner-lite.html">
 
     <style is="custom-style">
       .horizontal-section {
         width: 100px;
       }
 
-      paper-spinner, button {
+      paper-spinner, paper-spinner-lite, button {
```
Comment (cdata) [inline demo/index.html]: :dancers:

File: paper-spinner-behavior.html [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+-->
+
+<link rel="import" href="../polymer/polymer.html">
+
+<script>
+
+  /** @polymerBehavior */
+  Polymer.PaperSpinnerBehavior = {
```
Comment (cdata) [inline paper-spinner-behavior.html]: :+1:

File: paper-spinner-behavior.html [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    listeners: {
+      'animationend': 'reset',
+      'webkitAnimationEnd': 'reset'
+    },
+
+    properties: {
+      /**
+       * Displays the spinner.
+       *
+       * @attribute active
```
Comment (cdata) [inline paper-spinner-behavior.html]: These annotations are mostly unnecessary now. The compiler and Hydrolysis are both sufficiently robust that they can infer the values from the property declaration. The only time we need them now is if the documentation should be different from the declaration for some reason.

File: paper-spinner-behavior.html [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      /**
+       * True when the spinner is going from active to inactive. This is represented by a fade
+       * to 0% opacity to the user.
+       */
+      _coolingDown: {
+        type: Boolean,
+        value: false
+      },
+
+      _spinnerContainerClassName: {
```
Comment (cdata) [inline paper-spinner-behavior.html]: Conventionally, we document `protected` implementation. Anything prefixed with a single `_` is considered `protected`, while double (`__`) is considered `private`.

Comment (cdata) [inline paper-spinner-behavior.html]: FYI I'm going to leave some notes requesting documentation below. Feel free to document or convert to private implementation as you feel is appropriate.

File: paper-spinner-behavior.html [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        type: String,
+        computed: '_computeSpinnerContainerClassName(active, _coolingDown)'
+      }
+    },
+
+    _computeSpinnerContainerClassName: function(active, coolingDown) {
```
Comment (cdata) [inline paper-spinner-behavior.html]: Please document this method.

File: paper-spinner-behavior.html [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      return [
+        active || coolingDown ? 'active' : '',
+        coolingDown ? 'cooldown' : ''
+      ].join(' ');
+    },
+
+    _activeChanged: function(active, old) {
```
Comment (cdata) [inline paper-spinner-behavior.html]: Please document this method.

File: paper-spinner-behavior.html [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+      this._setAriaHidden(!active);
+      this._coolingDown = !active && old;
+    },
+
+    _altChanged: function(alt) {
```
Comment (cdata) [inline paper-spinner-behavior.html]: Please document this method.

File: paper-spinner-behavior.html [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+      // user-provided `aria-label` takes precedence over prototype default
+      if (alt === this.getPropertyInfo('alt').value) {
+        this.alt = this.getAttribute('aria-label') || alt;
+      } else {
+        this._setAriaHidden(alt==='');
+        this.setAttribute('aria-label', alt);
+      }
+    },
+
+    _setAriaHidden: function(hidden) {
```
Comment (cdata) [inline paper-spinner-behavior.html]: Please document this method.

File: paper-spinner-behavior.html [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+      var attr = 'aria-hidden';
+      if (hidden) {
+        this.setAttribute(attr, 'true');
+      } else {
+        this.removeAttribute(attr);
+      }
+    },
+
+    reset: function() {
```
Comment (cdata) [inline paper-spinner-behavior.html]: Please document this method.

File: paper-spinner-behavior.html [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        type: String,
+        computed: '_computeSpinnerContainerClassName(active, _coolingDown)'
+      }
+    },
+
+    _computeSpinnerContainerClassName: function(active, coolingDown) {
+      return [
+        active || coolingDown ? 'active' : '',
+        coolingDown ? 'cooldown' : ''
+      ].join(' ');
```
Comment (cdata) [inline paper-spinner-behavior.html]: We really need to create a `className` helper. I think @kevinpschaaf had something in the works for this. I pinged him about it. No action required, just observing.

File: paper-spinner-behavior.html [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+      } else {
+        this._setAriaHidden(alt==='');
+        this.setAttribute('aria-label', alt);
+      }
+    },
+
+    _setAriaHidden: function(hidden) {
+      var attr = 'aria-hidden';
+      if (hidden) {
+        this.setAttribute(attr, 'true');
```
Comment (cdata) [inline paper-spinner-behavior.html]: Are you willing to predict if the majority of usages will start with `aria-hidden` or without it in the spinner's default state? :currency_exchange:

**Comment by cdata** [comment]
This is looking really good :camel: :cactus:  thanks for making such a performance-conscious update to this element

**Comment by cdata** [comment]
I gave `paper-spinner-lite` a spin (heh) in the WIP benchmark element. Here is an example of what the test measures:

```js
 window.addEventListener('WebComponentsReady', function() {
  StartTest();
  var spinner = document.createElement('paper-spinner');
  spinner.active = true;
  document.body.appendChild(spinner);
  document.body.offsetHeight;
  EndTest();
});
```

And here is one of the results I got:

spinner-test

There is definitely an improvement, but given how much DOM you removed from `paper-spinner-lite`, I expected a bigger bump. WDYT?

File: paper-spinner-behavior.html (hunk 11 above)
Comment (keanulee) [inline paper-spinner-behavior.html]: I'm not sure if I understand the question. I think both `active` and not are likely initially depending on the use case (e.g. creating an inactive spinner in anticipation that it needs to be active later, or creating an active spinner for page load).

**Comment by keanulee** [comment]
:tada: Yay for perf! I suspect that at this point most of the cost comes from the JS/CSS animation initialization and there's not much that can be done.

Looking forward to this benchmark element - it's gonna be so helpful when evaluating these elements.

**Comment by abdonrd** [comment]
[Here](https://github.com/PolymerElements/paper-spinner/pull/41/files#diff-48612ea18dd8d1d06699b2156aae6120R53) and [here](https://github.com/PolymerElements/paper-spinner/pull/41/files#diff-da9be34d9d31e06d09ae905a6af1b03aR58), why create a new property instead of add the computed function here?

Changing this:
```html
<div id="spinnerContainer" class-name="[[__spinnerContainerClassName]]">
```

To:
```html
<div id="spinnerContainer" class-name="[[__computeSpinnerContainerClassName(active, __coolingDown)]]">
```

And then remove the [__spinnerContainerClassName](https://github.com/PolymerElements/paper-spinner/pull/41/files#diff-765ad7eb2e33e1b356c6c1924e08473cR50) property.

**Comment by ebidel** [comment]
Is there a reason we can't make this update to the existing element rather than creating a new one? Everyone already using paper-spinner could benefit from the perf tweaks!

**Comment by cdata** [comment]
@ebidel Thanks for bringing this up. The perf tweak comes at the cost of a major visual feature of the main element: transition between multiple colors. Yes, it is possible to implement this within the main element.

One approach might be to use `template is="dom-if"` to conditionally populate the local DOM with the needed nodes. However, this would most likely result in a performance penalty.

Another approach would be to do what we do with `PaperRippleBehavior`: imperatively create nodes as needed in an observer. However, given the relative density of DOM needed, this would be very unpleasant to write imperatively.

Still another option would be to break out of the realm of Polymer's conveniences and write code to conditionally stamp one plain template or another inside of an observer.

Ultimately, this performance improvement would have to be "opt-in" by anyone who wishes to benefit from it. The distinction between opting-in via a new element vs opting-in via an attribute is small. With this in mind, creating a new element strikes a nice balance between performance improvement and obtuseness of implementation.

**Comment by ebidel** [comment]
@cdata thanks for the explanation. Just trying to think of ways we decrease the surface of area of all our offerings. The more we have, the more bugs, maintenance, etc. I'll leave it to your judgement to decide which is the lesser of two evils :)

**Comment by keanulee** [comment]
@abdonrd Thanks for the tip - implemented.

**Comment by abdonrd** [comment]
@keanulee your welcome! :+1:

**Comment by cdata** [comment]
LGTM :+1:

**Pull request merged**